lib/ocpp/v201/smart_charging.cpp

```diff
@@ -315,12 +316,32 @@ SmartChargingHandler::validate_profile_schedules(ChargingProfile& profile,
     return ProfileValidationResultEnum::Valid;
 }
 
-void SmartChargingHandler::add_profile(int32_t evse_id, ChargingProfile& profile) {
-    if (STATION_WIDE_ID == evse_id) {
-        station_wide_charging_profiles.push_back(profile);
+SetChargingProfileResponse SmartChargingHandler::add_profile(int32_t evse_id, ChargingProfile& profile) {
+    SetChargingProfileResponse response;
+    response.status = ChargingProfileStatusEnum::Accepted;
+
+    auto& profile_storage = (STATION_WIDE_ID == evse_id) ? station_wide_charging_profiles : charging_profiles[evse_id];
```

Comment:
Why do we have a separate store for station wide profiles and evse related charging profiles? For this use case it should be a single list I think. This would also be more in line with how the database is going to look like.

K01.FR.05 says that a profile with the same id should replace the existing profile. There is no mention of if the new profile is for the same EVSE etc.

Reply:
> Why do we have a separate store for station wide profiles and evse related charging profiles? For this use case it should be a single list I think. This would also be more in line with how the database is going to look like.

I believe the separate `station_wide_charging_profiles` variable was created for clarity, since various expected behaviors involving profiles with EVSE ID 0 are different than for other EVSE IDs. (@couryrr-afs please correct me if I'm wrong here.)

Reply:
> K01.FR.05 says that a profile with the same id should replace the existing profile. There is no mention of if the new profile is for the same EVSE etc.

+1. If they're not already present, there should be test cases here for ensuring that adding a replacement charging profile (that is, one with the same ID as an existing profile) for a different EVSE than the existing profile results in the new profile being added and the old one being deleted.

Reply:
> > Why do we have a separate store for station wide profiles and evse related charging profiles? For this use case it should be a single list I think. This would also be more in line with how the database is going to look like.
> 
> I believe the separate `station_wide_charging_profiles` variable was created for clarity, since various expected behaviors involving profiles with EVSE ID 0 are different than for other EVSE IDs. (@couryrr-afs please correct me if I'm wrong here.)

I seem to recall a PR comment asking for them to be separated, although I am having trouble finding it now.

But yes, it was because the concept of a profile on EVSE with ID 0 is handled differently and so the code clarity of separating it explicitly made sense.

Reply:
K01.FR.05 says:

> When a SetChargingProfileRequest with an already known ChargingProfile.id is received AND the existing ChargingProfile does NOT have chargingProfilePurpose = `ChargingStationExternalConstraints`, The Charging Station SHALL replace the existing ChargingProfile with the one specified.

When [working on K01.FR.05](https://github.com/US-JOET/libocpp/pull/11) it wasn't clear whether or not the charging profile should be replaced when they're in different places. Station-wide profiles are treated differently from EVSE-specific profiles - they're meant to be apply to all EVSEs. K01.FR.14 also says "The Charging Station SHALL apply, but not copy, this profile to all EVSEs" for profiles on EVSE 0. So that implies that station-wide profiles should be independent from the EVSE-specific profiles. This made it seem odd to replace a station-wide profile that applies to all EVSEs with a profile that doesn't or vice-versa. Should those profiles be replaced then? If so, I think it makes sense to have a test case.

Reply:
As discussed in the working group call, requirement K01.FR.05 is pretty clear that the charging profile id should be unique for the whole charge point. Therefore it indeed should be replaced and I agree adding a test is the best course of action.

Reply:
We refactored how we store profiles and simplified the logic, and added new tests to ensure profile IDs are unique in all cases. We also added some new logic and tests to `validate_profile()` in order to ensure we don't replace `ChargingStationExternalConstraints` profiles.

Reply:
Looks like a good change, makes sense to use the map to also store the station wide profiles.